File: 00658. Find K Closest Elements/00658.c
// https://leetcode.com/problems/find-k-closest-elements

int x;

int compare(const void *aa, const void *bb) {
    int a = *(int *)aa, b = *(int *)bb;
    a = (a > x) ? a-x : x-a;
    b = (b > x) ? b-x : x-b;

    if (a == b) {
        return *(int *)aa - *(int *)bb;
    }
    return a - b;
}

int compare2(const void *aa, const void *bb) {
    return *(int *)aa - *(int *)bb;
}

int * findClosestElements(int *arr, int arrSize, int k, int target, int *returnSize) {
    x = target;
    *returnSize = k;
    int *ret = (int *) malloc(sizeof(int) * k);

    qsort(arr, arrSize, sizeof(int), compare);

    memcpy(ret, arr, sizeof(int) * k);
    qsort(ret, k, sizeof(int), compare2);

    return ret;
}

File: LeetCode/00658. Find K Closest Elements/00658.go
// https://leetcode.com/problems/find-k-closest-elements

func findClosestElements(arr []int, k int, x int) []int {
    size := len(arr)

    right := sort.SearchInts(arr, x)
    left := right-1
    
    if right == 0 {
        return arr[:k]
    } else if right == size {
        return arr[size-k:]
    }

    lDiff, rDiff := x - arr[left], arr[right] - x

    for k > 0 {
        k--

        if lDiff <= rDiff {
            left--
            if left < 0 {
                right += k
                break
            }

            lDiff = x - arr[left]
        } else {
            right++
            if right >= size {
                left -= k
                break
            }

            rDiff = arr[right] - x
        }
    }

    return arr[left+1:right]
}
